ejb/WFLY-17986.adoc

```diff
@@ -0,0 +1,234 @@
+
+Unless otherwise constrained, it would be very difficult for an EJB client application to interact with an EAP
+deployment in OpenShift if stricter guarantees were not imposed. For example, discovery of the members of a
+EAP cluster and affinity management of SFSBs would be next to impossible without an assumption persistent Pod
+IP addresses.
+
+The `WildFly operator` can be used to deploy an EAP application in OpenShift and provide the following guarantees:
+
+* persistent IP addresses for Pods
+* persistent storage references for storage used by EAP
+* scaling up or scaling down of services which are transaction-aware
+
+In what follows, we assume that deployments are made using the WildFly operator and that the EJB client may depend
+on these additional guarantees. For more information on the WildFly operator, see
+https://operatorhub.io/operator/WildFly[WildFly Operator on operatorhub.io]
+
+=== Scope
+As mentioned earlier, the scope of this issue is limited to `external` EJB clients using the `http` protocol to
+access `WildFly-operator-based` EAP deployments on OpenShift. Within this limited use case scenario, the aim is to
+provide the same set of features to the application programmer on OpenShift as are available on bare metal.
+
+== Issue Metadata
+
+=== Issue
+
+* https://issues.redhat.com/browse/WFLY-17986[WFLY-17986]
+
+=== Related Issues
+
+* https://issues.redhat.com/browse/EAP7-2062[EAP7-2062]
+
+=== Dev Contacts
+
+* mailto:{email}[{author}]
+
+=== QE Contacts
+
+=== Testing By
+* [ ] Engineering
+
+* [x] QE
+
+=== Affected Projects or Components
+
+* https://github.com/WildFly/WildFly[WildFly]
+* https://github.com/WildFly/WildFly-http-client[WildFly HTTP Client]
+
+=== Other Interested Projects
+
+* https://github.com/WildFly-security/WildFly-elytron[WildFly Elytron]
+
+=== Relevant Installation Types
+// Remove the x next to the relevant field if the feature in question is not relevant
+// to that kind of WildFly installation
+* [x] Traditional standalone server (unzipped or provisioned by Galleon)
+
+* [ ] Managed domain
+
+* [x] OpenShift s2i
+
+* [x] Bootable jar
+
+== Requirements
+The requirements for this issue center around the key features of the EJB client mentioned previously.
+We assume that the deployments are deployed using the WildFly operator.
+
+=== Hard Requirements
+
+* the key features of the EJB client should work on OpenShift as they do on bare metal, those features being:
+** lookup via JNDI or programmatic creation of proxies for EJBs in a deployment (non-HA and HA)
+** client-specific security and discovery configuration (non-HA and HA)
+** transactional scoping of invocations on EJBs in a deployment (non-HA and HA)
+** load balancing (HA only)
+** affinity management (HA only)
+** failover (HA only)
+
+* the affinity management feature should work with all load balancers supported by OpenShift; this includes:
```

Comment:
what is the mechanism that is going to provide such a feature? does it already exist?

Reply:
When using the EJB client with protocol http, the wildfly-http-client library is used to provide the transport of invocations from the client to the server. This transport has a client side part and a server side part and they work together to make sure that HTTP requests have their affinity managed correctly. The current implementation s broken and the issue is addressed by WEJBHTTP-81.